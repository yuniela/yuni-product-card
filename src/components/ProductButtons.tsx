import React, { useContext, CSSProperties, useCallback } from 'react';
import { ProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css';

export interface Props {
    className?: string;
    style?: CSSProperties;

}
export const ProductButtons = ({className, style}: Props ) => {
    //Todo: maxCount
    const { increaseBy, counter, maxCount} = useContext( ProductContext);

    //Todo: isMaxReached = useCallback, [count, maxCount]
    // TRUE si el count === maxCount
    const isMaxReached = useCallback(
        () => !!maxCount && counter === maxCount, 
        [counter, maxCount]
    )

    return (
        <div 
            className={ `${styles.buttonsContainer} ${ className }`}
            style= { style }
        >
        <button 
            className={ styles.buttonMinus }
            onClick={ () => increaseBy( -1 )}
        >-</button>
        <div className={ styles.countLabel}>{ counter }</div>
        <button 
            className={ `${ styles.buttonAdd } ${ isMaxReached() && styles.disabled }` }
            onClick={ () => increaseBy( 1 )}
        >+</button>
      </div>
    )
}