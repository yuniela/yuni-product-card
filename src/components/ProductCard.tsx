import styles from '../styles/styles.module.css';
import React, { createContext, CSSProperties } from 'react';
import useProduct from '../hooks/useProduct';
import { ProductContextProps, Product, onChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';
import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';


export const ProductContext = createContext({} as ProductContextProps);

const { Provider } = ProductContext;

export interface Props {
  product: Product;
  // children?: ReactElement | ReactElement[];
  children: ( args: ProductCardHandlers ) => JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: ( args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({ children, product, className, style, onChange, value, initialValues }: Props) => {

  const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });
    
  return (
    <Provider value={{
        counter,
        increaseBy,
        maxCount,
        product
    }}>
        <div 
          className={ `${styles.productCard} ${className}` }
          style={ style }
        >
            { children({
              count: counter,
              isMaxCountReached,
              maxCount: initialValues?.maxCount,
              product,
              increaseBy,
              reset
            }) }
        </div>
    </Provider>
  )
}

ProductCard.Title  = ProductTitle;
ProductCard.Image   = ProductImage;
ProductCard.Buttons = ProductButtons;
