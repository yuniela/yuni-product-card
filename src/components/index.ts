import { ProductCard as ProductCardHOC } from './ProductCard';

import { ProductTitle } from './ProductTitle';
import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProductCardHOCProps } from '../interfaces/interfaces';

export { ProductTitle } from './ProductTitle';
export { ProductButtons } from './ProductButtons';
export { ProductImage } from './ProductImage';


export const ProductCard: ProductCardHOCProps = Object.assign( ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons
})

export default ProductCard;