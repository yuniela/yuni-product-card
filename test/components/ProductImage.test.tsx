import React from "react";
import renderer from 'react-test-renderer';
import { ProductImage, ProductCard } from '../../src/components';
import { product2 } from '../data/products';

describe('ProductImage', () => {
    test('Should render an image', () => {
        const wrapper = renderer.create(
            <ProductImage img="https://hola.jpg"/>
        )

        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    test('Should render a component with a product name', () => {
        const wrapper = renderer.create(
            <ProductCard product={product2}>
                {
                    () => (
                        <ProductImage/>
                    )
                }

            </ProductCard>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
    })
})