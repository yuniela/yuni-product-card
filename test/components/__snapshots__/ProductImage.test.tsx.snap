// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Should render a component with a product name 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`ProductImage Should render an image 1`] = `
<img
  alt="Product"
  className="productImg undefined"
  src="https://hola.jpg"
/>
`;
